### PR TITLE
ref(ts): converts plugin registry

### PR DESCRIPTION
**Pull request by scefali**
I've converted the plugin registry file along with some of the base plugin classes.

In `Registry`, I removed some functionality:
1. Removed the unused`loadAll` function
2. Removed asset loading. No plugins in SaaS Sentry are using this feature. It's technically possible for on-prem users to use this functionality with community plugins but it seems unlikely. Even if there were, I don't think we should maintain this feature since it's kind of a janky solution.

### CODE REVIEW

File: src/sentry/static/sentry/app/plugins/index.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import SessionStackPlugin from './sessionstack';
 import SessionStackContextType from './sessionstack/contexts/sessionstack';
 import Jira from './jira';
 
-const contexts = {};
+const contexts: Record<string, any> = {};
```
Comment (billyvg) [inline src/sentry/static/sentry/app/plugins/index.tsx:9]: Can we be specific for the value? are they always `React.ElementType`?

File: src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx [hunk 2]
```diff
@@ -1,14 +0,0 @@
...
-import BasePlugin from 'app/plugins/basePlugin';
-import IssueActions from 'app/plugins/components/issueActions';
-
-export class DefaultIssuePlugin extends BasePlugin {
-  renderGroupActions(props) {
-    return <IssueActions plugin={this} {...props} />;
-  }
-}
-
-DefaultIssuePlugin.DefaultIssueActions = IssueActions;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx:12]: Will removing this break anything?

File: src/sentry/static/sentry/app/plugins/basePlugin.jsx [hunk 3]
```diff
@@ -1,17 +0,0 @@
...
-  constructor(data) {
-    Object.assign(this, data);
-  }
-
-  renderSettings(props) {
-    return <Settings plugin={this} {...props} />;
-  }
-}
-
-BasePlugin.DefaultSettings = Settings;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/plugins/basePlugin.jsx:15]: Will removing this break anything?

File: src/sentry/static/sentry/app/plugins/jira/index.jsx [hunk 4]
```diff
@@ -7,11 +7,11 @@ import IssueActions from './components/issueActions';
 
 class Jira extends DefaultIssuePlugin {
   renderSettings(props) {
-    return <Settings plugin={this} {...props} />;
+    return <Settings plugin={this.plugin} {...props} />;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/plugins/jira/index.jsx:10]: is `this` == `this.plugin`?

File: src/sentry/static/sentry/app/plugins/basePlugin.jsx (hunk 3 above)
Comment (markstory) [reply src/sentry/static/sentry/app/plugins/basePlugin.jsx:15]: Looks like both uses of this were also updated.

File: src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx (hunk 2 above)
Comment (scefali) [reply src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx:12]: @billyvg not with the refactor that I did as Mark points out

File: src/sentry/static/sentry/app/plugins/jira/index.jsx (hunk 4 above)
Comment (scefali) [reply src/sentry/static/sentry/app/plugins/jira/index.jsx:10]: @billyvg It is after this change:
```
  constructor(data: Plugin) {
    this.plugin = data;
  }
```

File: src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx (hunk 2 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx:12]: @scefali Are these modules used for external/custom plugins? could we break those?

Comment (scefali) [reply src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx:12]: @billyvg I searched on Github and I don't think anyone is hooking into plugins this way. Most plugins don't use these modules. I would be surprised if any community-based plugins use this. I don't think it's something we should support long term as well.

Comment (scefali) [reply src/sentry/static/sentry/app/plugins/defaultIssuePlugin.jsx:12]: By most, I mean all but Jira and SessionStack don't use this hook

**Review by billyvg — APPROVED** [review]

**Comment by nikhilkalige** [comment]
@scefali I am trying to add a plugin for our onprem installation which loads a javascript asset. With the removal of the api in `registry.jsx`, only way would be to have a fork or use nginx to insert that script. So, is there a better way to inject assets, or is there a plan to provide a similar feature in a different way. 

Thanks.

**Comment by scefali** [comment]
@nikhilkalige I didn't realize anyone was using this system. I will look into adding it back.

**Pull request merged**